packages/sanes-chrome-extension/src/extension/background/externalHandler/index.ts

```diff
@@ -0,0 +1,92 @@
+/*global chrome*/
+import {
+  JsonRpcErrorResponse,
+  JsonRpcResponse,
+  jsonRpcCode,
+  parseJsonRpcRequest,
+  JsonRpcRequest,
+} from '@iov/jsonrpc';
+import { UseOnlyJsonRpcSigningServer } from '../../../logic/persona';
+import { RequestHandler, parseMethod } from './requestHandler';
+import { SenderWhitelist } from './senderWhitelist';
+
+type SigningServer = UseOnlyJsonRpcSigningServer | undefined;
+
+function generateErrorResponse(id: number | null, message: string): JsonRpcErrorResponse {
+  return {
+    jsonrpc: '2.0',
+    id,
+    error: {
+      code: jsonRpcCode.serverError.default,
+      message,
+    },
+  };
+}
+
+function updateExtensionBadge(): void {
+  const isExtensionContext = typeof chrome !== 'undefined'; // needed for tests
+  if (!isExtensionContext) {
+    return;
+  }
+
+  const queueSize = RequestHandler.requests().length;
+  const badgeText = queueSize === 0 ? '' : `${queueSize}`;
+  const iconPath = queueSize === 0 ? 'assets/icons/icon128.png' : 'assets/icons/request128.png';
+
+  chrome.browserAction.setIcon({ path: iconPath });
+  chrome.browserAction.setBadgeBackgroundColor({ color: [0, 0, 0, 0] });
+  chrome.browserAction.setBadgeText({ text: badgeText });
+}
+
+export async function handleExternalMessage(
+  signingServer: SigningServer,
+  externalRequest: any, // eslint-disable-line @typescript-eslint/no-explicit-any
+  sender: string | undefined,
+): Promise<JsonRpcResponse> {
+  const responseId = typeof externalRequest.id === 'number' ? externalRequest.id : null;
+  if (!sender) {
+    return generateErrorResponse(responseId, 'Unkown sender, droped request');
+  }
+
+  if (!signingServer) {
+    return generateErrorResponse(responseId, 'Signing server not ready');
+  }
+
+  if (!SenderWhitelist.loaded()) {
+    return generateErrorResponse(responseId, 'SenderWhitelist has not been initialised');
+  }
+
+  let request: JsonRpcRequest;
+  try {
+    request = parseJsonRpcRequest(externalRequest);
+    parseMethod(request.method);
+  } catch (err) {
+    return generateErrorResponse(responseId, err.message);
+  }
+
+  if (SenderWhitelist.isBlocked(sender)) {
+    return generateErrorResponse(responseId, 'Sender has been blocked by user');
+  }
+
+  return new Promise(resolve => {
```

Comment:
All the queueing should happen at the level of autorization requests. That is the point where the signing server asks the UI/user: _please handle the request and tell me what I should do with it_. Since the UI is not always available, those requests can be queued.

To queue pending callbacks, you can do:

```ts
interface QueuedElement {
  reason: string;
  senderUrl: string;
  accept(): void;
  reject(): void;
}

const queue = new Array<QueuedElement>();

const getIdentitiesCallback: GetIdentitiesAuthorization = async (
  _reason,
  matchingIdentities,
): Promise<ReadonlyArray<PublicIdentity>> => {
  // sender will be available here after upgrading to IOV-Core 0.13.7
  // https://github.com/iov-one/iov-core/pull/993
  if (isNewSender('http://finnex.com')) {
    chrome.browserAction.setIcon({ path: 'assets/icons/request128.png' });
    chrome.browserAction.setBadgeBackgroundColor({ color: [0, 0, 0, 0] });
    chrome.browserAction.setBadgeText({ text: '*' });
  }

  const pendingGetIdentititesRequest = new Promise<ReadonlyArray<PublicIdentity>>(resolve => {
    function acceptThisRequest(): void {
      resolve(matchingIdentities); // select all
    }

    function rejectThisRequest(): void {
      resolve([]); // select none
    }

    queue.push({
      reason: _reason,
      senderUrl: 'http://finnex.com',
      accept: acceptThisRequest,
      reject: rejectThisRequest,
    });
  });

  return pendingGetIdentititesRequest;
};
```

Reply:
I can create it in the internalHandler when I create the persona and therefore restart the RequestHandler easily. I do not see a blocker on this. So far so good.

What about signAndPost? That was my main problem, I receive the callback once the tx is signed, how can I reject it?

Reply:
> What about signAndPost? That was my main problem, I receive the callback once the tx is signed, how can I reject it?

You receive the callback before the transaction is signed. You can click on the types to get a bunch of docs:

```ts
export interface SignAndPostAuthorization {
    /**
     * Called by the signing server and lets user decide if they want to
     * authorize a sign and post request.
     *
     * Must return true if the user authorized the signing and false if the
     * user rejects it.
     *
     * Any error (thrown or returned as a rejected promise) from this
     * callback is sent to the RPC client as an unspecified
     * "Internal server error" and the callback author should
     * ensure this does not happen.
     *
     * @param reason An explanation why the autorization is requested. This is created by the website and shown to the user.
     * @param transaction The transaction to be signed.
     * @param meta An object that is passed by reference from request handlers into the callback.
     */
    (reason: string, transaction: UnsignedTransaction, meta?: any): Promise<boolean>;
}
```